### ISSUE TITLE
Failure to detect non-None Optional type in conditional expression

### ISSUE THREAD

**Issue by jonafato** [python/mypy#6755]
* Are you reporting a bug, or opening a feature request?
Bug

* Please insert below the code you are checking with mypy, or a mock-up repro if the source is private. We would appreciate if you try to simplify your case to a minimal repro.
```python
from typing import Mapping, Optional

names_to_numbers: Mapping[str, Optional[int]] = {
    'one': 1,
    'two': 2,
    'foo': None,
}

squares = {
    1: 1,
    2: 4,
    3: 9,
}

squares[names_to_numbers['foo']] if names_to_numbers['foo'] else None
```

* What is the actual behavior/output?
```
bug.py:15: error: Invalid index type "Optional[int]" for "Dict[int, int]"; expected type "int"
```
This error is only thrown for conditional expressions of the form `if container[index] else None`. Changing it to any of the following works as expected:

```python
squares[names_to_numbers['foo']] if names_to_numbers['foo'] is not None else None
```
```python
squares[names_to_numbers['foo']] if isinstance(names_to_numbers['foo'], int) else None
```
```python
foo = names_to_numbers['foo']
squares[foo] if foo else None
```

* What is the behavior/output you expect?
No errors reported.

* What are the versions of mypy and Python you are using?
mypy 0.701, reproducible on at least Python 2.7 and 3.7.

*  Do you see the same issue after installing mypy from Git master?
Yes

**Comment by ilevkivskyi** [comment]
I agree this behavior is inconsistent. Mypy already allows (on purpose) some code that may be technically unsafe (e.g. `__getitem__()` call is not guaranteed to return same return with same arguments every time) but are practically useful. We already allow this in other circumstances as you noticed. We should probably to update the branch in `find_isinstance_check()` (not the best name, I know) where we test for `if isinstance(node, RefExpr)` to accept any `BindableExpression` with a correct literal level.

Btw, the same same happens in normal if statement, not only in a ternary expression:
```python
if names_to_numbers['foo']:  # error also disappears if I add 'is not None'
    squares[names_to_numbers['foo']]  # same error here
```

**Comment by mbdevpl** [comment]
I'm wondering is this issue essentially the same as what I'm getting, or should I submit a new one?

```console
$ cat conditions.py
```

```python3
def conditional(condition: bool = False) -> tuple:
    if condition:
        return ((1, '', 0),)
    return ()


def inline_conditional(condition: bool = False) -> tuple:
    return ((1, '', 0),) if condition else ()  # line 8 is here
```

```console
$ mypy conditions.py
conditions.py:8: error: Incompatible return value type (got "object", expected "Tuple[Any, ...]")
Found 1 error in 1 file (checked 1 source file)
```

My environment:

* Ubuntu 19.10
* Python 3.8.0 (default, Oct 23 2019, 16:52:33) [GCC 9.2.1 20191008]
* mypy 0.761
* typed-ast 1.4.1

I know this issue is about Optional types, but from my observations, mypy fails to detect multitude of types when they are returned for the ternary conditional operator, not just Optional.

**Comment by mbdevpl** [comment]
And yeah, I forgot to mention (which was the point of the example) that after rewriting the ternary conditional operator to normal if statement, the types are inferred correctly.

**Comment by JelleZijlstra** [comment]
@mbdevpl that's a separate issue with the ternary expression, similar to #4134.

**Comment by mbdevpl** [comment]
@JelleZijlstra OK, migrated there! Thanks!

**Comment by AlexWaygood** [comment]
This is [no longer reproducible](https://mypy-play.net/?mypy=0.930&python=3.10&flags=strict&gist=99a977a5b92a482c39f31fb414c3aa74) on mypy 0.930+
